packages/Search/src/details/DTK_DetailsDistributor.hpp

```diff
@@ -0,0 +1,252 @@
+        "" );
+    static_assert( std::is_same<typename OutputView::value_type, int>::value,
+                   "" );
+    static_assert(
+        Kokkos::Impl::MemorySpaceAccess<typename OutputView::memory_space,
+                                        Kokkos::HostSpace>::accessible,
+        "" );
+
+    offsets.push_back( 0 );
+
+    auto const n = ranks.extent_int( 0 );
+    if ( n == 0 )
+        return;
+
+    Kokkos::View<int *, Kokkos::HostSpace> ranks_duplicate(
+        Kokkos::ViewAllocateWithoutInitializing( ranks.label() ),
+        ranks.size() );
+    Kokkos::deep_copy( ranks_duplicate, ranks );
+
+    while ( true )
+    {
+        // TODO consider replacing with parallel reduce
+        int const largest_rank = *std::max_element(
+            ranks_duplicate.data(), ranks_duplicate.data() + n );
+        if ( largest_rank == -1 )
+            break;
+        unique_ranks.push_back( largest_rank );
+        counts.push_back( 0 );
+        // TODO consider replacing with parallel scan
+        for ( int i = 0; i < n; ++i )
+        {
+            if ( ranks_duplicate( i ) == largest_rank )
+            {
+                ranks_duplicate( i ) = -1;
+                permutation_indices( i ) = offsets.back() + counts.back();
+                ++counts.back();
+            }
+        }
+        offsets.push_back( offsets.back() + counts.back() );
+    }
+}
+
+// A number of MPI routines (in the current OpenMPI implementation at least) are
+// checking at runtime that all pointers passed to them as argument are not null
+// and call program termination even when no data would actually be
+// communicated. This is problematic when using std::vector to store buffers and
+// other arrays because std::vector::data() may return a null pointer if the
+// vector is empty. This function "swaps" the pointer with some other pointer
+// that is not dereferenceable if it is the null pointer.
+template <typename T>
+static T *notNullPtr( T *p )
+{
+    return p ? p : reinterpret_cast<T *>( -1 );
+}
+
+class Distributor
+{
+  public:
+    Distributor( MPI_Comm comm )
+        : _comm( comm )
+    {
+    }
+
+    template <typename View>
+    size_t createFromSends( View const &destination_ranks )
+    {
+        static_assert( View::rank == 1, "" );
+        static_assert(
+            std::is_same<typename View::non_const_value_type, int>::value, "" );
+        int comm_rank;
+        MPI_Comm_rank( _comm, &comm_rank );
+
+        _permute = Kokkos::View<int *, Kokkos::HostSpace>(
+            Kokkos::ViewAllocateWithoutInitializing( "permute" ),
+            destination_ranks.size() );
+        std::vector<int> destinations;
+        sortAndDetermineBufferLayout( destination_ranks, _permute, destinations,
+                                      _dest_counts, _dest_offsets );
+
+        {
+            int const reorder = 0; // Ranking may *not* be reordered
+            int const sources[1] = {comm_rank};
```

Comment:
> So you only have one source in this call to `MPI_Dist_graph_create` because we only know where this rank is sending right?

Correct

> I wonder if it will resolve collisions where the same edge is input on multiple ranks.

It is ok to list the same edge on different processes.  It will only appears once in the resulting communication graph.  I am not sure how it would handle weights in that case though.